nservicebus/hosting/critical-errors.md

```diff
@@ -31,11 +31,18 @@ Define a custom handler using the following code.
 
 snippet: DefiningCustomHostErrorHandlingAction
 
-### Terminating the process
+### Stopping the endpoint
 
-It is often unknown if the issue is recoverable when a critical error occurs. A sound strategy is to terminate the process when a critical error occurs and rely on the process hosting environment to restart the process as a recovery mechanism, resulting in a resilient way to deal with critical errors.
+It is often unknown if the issue is recoverable when a critical error occurs. A sound strategy is to terminate the process (e.g. via `Environment.FailFast` or `IHostApplicationLifetime.Stop`) when a critical error occurs and rely on the process hosting environment to restart the process as a recovery mechanism, resulting in a resilient way to deal with critical errors.
+
+Before terminating the process, the NServiceBus endpoint can attempt a graceful shutdown:
+
+snippet: StopEndpointInCriticalError
+
+Note: [The Microsoft Generic Host's](/nservicebus/hosting/extensions-hosting.md) `IHostApplicationLifetime.Stop` method also stops the NServiceBus endpoint gracefully.
```

Comment:
```suggestion
Note: When using [NServiceBus.Extensions.Hosting](/nservicebus/hosting/extensions-hosting.md) the Microsoft Generic Host's `IHostApplicationLifetime.Stop` method also stops the NServiceBus endpoint gracefully.
```

Reply:
Also not sure if this should be here. Maybe its better to add this to https://docs.particular.net/nservicebus/hosting/extensions-hosting and link to it as something like. Please review hosting packages for behavior on critical errors and stop and/or shutdown behavior?

Reply:
I was thinking about this but this is specifically about how to stop the endpoint/process as part of a critical error and might be a very common error to run into. Not against adding some documentation to extension hosting about how to properly shutdown the endpoint but I think that's a separate PR.